src/main/java/hudson/plugins/active_directory/HttpHeaderFilter.java

```diff
@@ -0,0 +1,127 @@
+
+                    UsernamePasswordAuthenticationToken auth = new UsernamePasswordAuthenticationToken(userName, "", authorities);
+
+                    try (ACLContext ignore = ACL.as(auth)) {
+                        chain.doFilter(request, response);
+                    }
+                    return;
+                } catch (UsernameNotFoundException e) {
+                    LOGGER.log(Level.FINE, "User from HTTP Header {0} not found in LDAP", userName);
+                }
+            }
+        }
+        chain.doFilter(request, response);
+    }
+
+    String getUserFromAuthorizationHeader(HttpServletRequest request) {
+        String authorization;
+        if ((authorization = request.getHeader("Authorization")) != null && authorization.toLowerCase().startsWith("basic ")) {
+            String uidpassword = Scrambler.descramble(authorization.substring(6));
+            int idx = uidpassword.indexOf(':');
+            if (idx >= 0) {
+                String username = uidpassword.substring(0, idx);
+                String password = uidpassword.substring(idx + 1);
+
+                // attempt to authenticate as API token
+                User u = User.get(username, false, Collections.emptyMap());
+                if (u != null) {
+                    ApiTokenProperty t = u.getProperty(ApiTokenProperty.class);
+                    if (t != null && t.matchesPassword(password)) { // Authenticate API-Token
+                        return username;
+                    } else { // Authenticate against LDAP
+                        UsernamePasswordAuthenticationToken auth = new UsernamePasswordAuthenticationToken(username, password);
+                        UserDetails details = activeDirectorySecurityRealm.getAuthenticationProvider().retrieveUser(username, auth);
+                        if (details != null)
+                            return username;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    String getUserFromReverseProxyHeader(HttpServletRequest request) {
+        String userFromHeader;
+        if ((getUserHeader() != null && (userFromHeader = request.getHeader(getUserHeader())) != null)) {
+            LOGGER.log(Level.FINE, "User from HTTP Header: {0}", userFromHeader);
+            if (getUsernameExtractionExpression() != null && !getUsernameExtractionExpression().toString().equals("")) {
+                Matcher m = getUsernameExtractionExpression().matcher(userFromHeader);
+                if (m.find() && m.groupCount()==1) {
+                    return m.group(1);
+                } else
+                    return userFromHeader;
+            } else
+                return userFromHeader;
+        }
+        LOGGER.log(Level.FINE, "NO user header found");
+        return null;
+    }
+
+    private String getUserHeader() {
+        return activeDirectorySecurityRealm.userFromHttpHeader;
```

Comment:
if this is the emtpy string how does `HttpServletRequest` request perform?

Reply:
from https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeader(java.lang.String)
>  If the request did not include a header of the specified name, this method returns null. 

This case is covered in line 95

Reply:
I am not talking about the header not being present - I am talking about `userFromHttpHeader` = `""`  
thus what does `HttpServletRequest.getHeader("")`  return?  

it is certainly undefined at the javadoc level and would be in violation of the HTTP spec - so an implementation could potentially throw an arbitrary `RuntimeException` here?

Reply:
The running implementation of `HttpServletRequest.getHeader` is interchangeable. 
I've checked it in `jetty-9.4.30.v20200611`:
 1. https://github.com/eclipse/jetty.project/blob/jetty-9.4.30.v20200611/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java#L852
 2. https://github.com/eclipse/jetty.project/blob/jetty-9.4.30.v20200611/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java#L270-L279

I don't see the potential for a `RuntimeException`. Either there is a HTTP-Header that matches `""` than it is returned or there is no match. That should yield `null`. Both cases are handled.